File: solutions/js/11.js
/**
 * @param {string[]} warehouse - The warehouse layout
 * @returns {number} The count of unwatched gifts
 */
function findUnsafeGifts(warehouse) {
  let unsafeCount = 0;
  for (let i = 0; i < warehouse.length; i++) {
    const row = warehouse[i];
    for (let j = 0; j < row.length; j++) {
      const item = warehouse[i][j];
      if (item !== "*") continue;
      const upperItem = warehouse[i - 1]?.[j];
      const bottomItem = warehouse[i + 1]?.[j];
      const leftItem = warehouse[i][j - 1];
      const rightItem = warehouse[i][j + 1];
      const validateArray = [upperItem, bottomItem, leftItem, rightItem];
      if (!validateArray.includes("#")) unsafeCount++;
    }
  }
  return unsafeCount;
}

const unsafe1 = findUnsafeGifts([".*.", "*#*", ".*."]); // ➞ 0
// All the presents are next to a camera
console.log(unsafe1);

const unsafe2 = findUnsafeGifts(["...", ".*.", "..."]); // ➞ 1
// This present has no cameras around it
console.log(unsafe2);

const unsafe3 = findUnsafeGifts(["*.*", "...", "*#*"]); // ➞ 2
// The presents in the top corners have no cameras around them
console.log(unsafe3);

const unsafe4 = findUnsafeGifts([".....", ".*.*.", "..#..", ".*.*.", "....."]); // ➞ 4
// The four presents have no cameras, because they are diagonal to the camera
console.log(unsafe4);

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 230
  ops: 3018
*/

File: solutions/python/11.py
def find_unsafe_gifts(warehouse: list[str]) -> int:
    rows = len(warehouse)
    cols = len(warehouse[0])
    unsafe_count = 0

    for i, row in enumerate(warehouse):
        for j, char in enumerate(row):
            if char != '*':
                continue
            if not any(
                0 <= ni < rows and 0 <= nj < cols and warehouse[ni][nj] == '#'
                for ni, nj in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]
            ):
                unsafe_count += 1

    return unsafe_count


unsafe1 = find_unsafe_gifts([".*.", "*#*", ".*."])  # ➞ 0
# All the presents are next to a camera
print(unsafe1)

unsafe2 = find_unsafe_gifts(["...", ".*.", "..."])  # ➞ 1
# This present has no cameras around it
print(unsafe2)

unsafe3 = find_unsafe_gifts(["*.*", "...", "*#*"])  # ➞ 2
# The presents in the top corners have no cameras around them
print(unsafe3)

unsafe4 = find_unsafe_gifts(
    [".....", ".*.*.", "..#..", ".*.*.", "....."])  # ➞ 4
# The four presents have no cameras, because they are diagonal to the camera
print(unsafe4)

"""
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 80
  ops: 2524
"""
